controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java

```diff
@@ -327,15 +344,20 @@ private String getWritersTableName(String id) {
     }
 
     @Override
-    public CompletableFuture<Void> removeSubscriber(String subscriber) {
+    public CompletableFuture<Void> removeSubscriber(String subscriber, long generation) {
         return Futures.toVoid(getSubscriberSetRecord(true)
-                .thenCompose(subscriberSetRecord -> getMetadataTable().thenCompose(table -> {
-                    SubscriberSet subSet = SubscriberSet.remove(subscriberSetRecord.getObject(), subscriber);
-                    return storeHelper.updateEntry(table, SUBSCRIBER_SET_KEY, subSet.toBytes(), subscriberSetRecord.getVersion())
-                            .thenCompose(v -> getSubscriberRecord(subscriber)
-                            .thenCompose(subscriberRecord -> storeHelper.removeEntry(table, getKeyForSubscriber(subscriber)))
-                            .thenAccept(x -> storeHelper.invalidateCache(table, SUBSCRIBER_SET_KEY))
-                            .thenAccept(x -> storeHelper.invalidateCache(table, getKeyForSubscriber(subscriber))));
+                .thenCompose(subscriberSetRecord -> getMetadataTable()
+                .thenCompose(table -> {
+                    if (subscriberSetRecord.getObject().getSubscribers().containsKey(subscriber)
+                        && subscriberSetRecord.getObject().getSubscribers().get(subscriber).longValue() < generation) {
+                        Subscribers subSet = Subscribers.remove(subscriberSetRecord.getObject(), subscriber);
+                        return getSubscriberRecord(subscriber)
+                                .thenCompose(subscriberRecord -> storeHelper.removeEntry(table, getKeyForSubscriber(subscriber)))
```

Comment:
this is not done atomically with updateSubscribers. 
so you have two steps:
updateSubscribers + removeSubscriberEntry
the problem is after updateSubscribers if someone added the subscriber back and created the subscriber entry idempotently, you may end up removing it.

Reply:
A solution could be do to conditional remove.